[Array API] Add linalg.vecdot

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70542

This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

lezcano committed:
[Array API] Add linalg.vecdot

This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

[ghstack-poisoned]

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -2251,3 +2251,42 @@
+vecdot = _add_docstr(_linalg.linalg_vecdot, r"""
+linalg.vecdot(x, y, *, dim=-1, out=None) -> Tensor
+
+Computes the vector product of two batches of vectors along a dimension.
```
There are several "vector products", this one computes the vector _dot_ product

Lezcano replied:
Yup. The only thing is that this PR implements \sum x_i y_i, which is not the complex dot product. I opened an issue here https://github.com/data-apis/array-api/issues/356. For now, I'm calling this "the real dot product", which is not standard notation either.

I don't think that there's a standard name for this in the complex case. You'd call it C-linear complex dot product (perhaps) to differentiate it from the R-linear real dot product `Re(x^H y)`...

Leaving it like this for now.

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -2251,3 +2251,42 @@
+
+over the dimension :attr:`dim`.
+
+Supports input of half, bfloat16, float, double, cfloat, cdouble and integral dtypes.
```
If it supports everything do we need to mention it?

I think it's fine to be explicit here. Also, many other linalg functions have a note on supported dtypes.

Lezcano replied:
Leaving it as "better explicit than implicit". Also, it does support `bool`, but it's not announced here. If I had to choose I'd disallow integral and boolean types, but left them because why not.

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -2251,3 +2251,42 @@
+    y (Tensor): second batch of vectors.
+
+Keyword args:
+    dim (int): Dimension along which to compute the vector product.
```
```suggestion
    dim (int): Dimension along which to compute the vector product. Default: `-1`.
```

IvanYashchuk reviewed (approved):
Great!

This PR also resolves https://github.com/pytorch/pytorch/issues/18027.

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

mruberry commented on torch/_torch_docs.py:
```diff
@@ -3429,21 +3429,34 @@ def merge_dicts(*dicts):
 
+.. seealso::
+
+        :func:`torch.linalg.vecdot` computes the dot product of two batches of vectors along a dimension.
```
Nice seealso addition

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -2251,3 +2251,38 @@
+It also supports broadcasting.
+
+Args:
+    x (Tensor): first batch of vectors.
```
If this supports batching should the main doc be updated with a sentence on batching, too?

mruberry replied:
(It's mentioned briefly in the initial sentence describing the operation, but like, what output shape should the user expect? Is it obvious?)

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1101,7 +1101,7 @@ class ReductionOpInfo(OpInfo):
 
     If a reduction operator does not yet implement the full required API of
-    reduction operators, this should be documented by skipping the failing
+    reduction operators, this should be documented by xfailing the failing
```
Nice update

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2012,6 +2012,15 @@ def sample_inputs_isclose(
                           kwargs=dict(op_kwargs, rtol=rtol, atol=atol, equal_nan=equal_nan))
 
+def sample_inputs_linalg_vecdot(op_info, device, dtype, requires_grad, **kwargs):
+    yield from sample_inputs_binary_pwise(op_info, device, dtype, requires_grad)
```
Not in love with reusing the elementwise binary sample inputs for this op -- wouldn't it be simple to generate some examples with and without batch dims, an empty tensor, and dim of the default, -2, 1, and 0?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9778,6 +9787,16 @@ def ref_pairwise_distance(input1, input2):
            ),
+    OpInfo('linalg.vecdot',
+           aten_name='linalg_vecdot',
+           ref=lambda x, y, *, dim=-1: (x.conj() * y).sum(dim),
```
Nice reference

mruberry reviewed (commented):
Overall looks great, @Lezcano. I left a few inline comments -- the one change I think we want to make is to just write a few sample inputs for this op specifically, especially because it doesn't have any tests beyond OpInfo tests. Just ping me when this ready to merge

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

Lezcano commented:
This one should be ready @mruberry.

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

anjali411 commented:
@pytorchbot rebase

Lezcano commented:
Yes, we should land this one. It was a bit left out due to the whole primtorch effort, but we should still land it.

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

Lezcano commented:
I just rebased the PR. If the tests go green, I'd say we merge it.

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

mruberry reviewed (approved):
Ship it!

Lezcano commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "Broke CUDA-10.2 for vecdot_bfloat16, see https://hud.pytorch.org/pytorch/pytorch/commit/74208a9c68b5892b9dde39d06350fe7b92691429" -c nosignal

malfet commented:
@Lezcano linalg PRs often cause CUDA-10.2 regressions, can you please test for it before submitting changes?

Lezcano commented:
The CI was green, so I assumed everything was fine. Should I run PRs on `trunk` or how should I test them? @malfet

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

lezcano committed:
Update on "[Array API] Add linalg.vecdot"


This PR adds the function `linalg.vecdot` specified by the [Array
API](https://data-apis.org/array-api/latest/API_specification/linear_algebra_functions.html#function-vecdot)

For the complex case, it chooses to implement \sum x_i y_i. See the
discussion in https://github.com/data-apis/array-api/issues/356

Edit. When it comes to testing, this function is not quite a binopt, nor a reduction opt. As such, we're this close to be able to get the extra testing, but we don't quite make it. Now, it's such a simple op that I think we'll make it without this.

Resolves https://github.com/pytorch/pytorch/issues/18027.

cc mruberry rgommers pmeier asmeurer leofang AnirudhDagar asi1024 emcastillo kmaehashi

[ghstack-poisoned]

Lezcano commented:
@pytorchbot merge -g
